Fast iterative Feature adversaries

fartashf opened the pull request.

goodfeli commented on examples/imagenet_featadvs/attack_model_featadv.py:
```diff
@@ -0,0 +1,160 @@
+"""Runs CleverHans attacks on the Madry Lab MNIST challenge model
```
If this is for the MNIST challenge, why is it in a directory with "imagenet" in the name?

fartashf replied:
Fixed in 47b04aa.

goodfeli commented on examples/imagenet_featadvs/attack_model_featadv.py:
```diff
@@ -0,0 +1,160 @@
+import numpy as np
+
+import tensorflow as tf
+# from tensorflow.examples.tutorials.mnist import input_data
```
Let's not commit commented code

fartashf replied:
Fixed in 47b04aa.

goodfeli commented on examples/imagenet_featadvs/attack_model_featadv.py:
```diff
@@ -0,0 +1,160 @@
+    # change the automatic naming of the variables.
+    model = MadryMNIST()
+
+    # x_input = tf.placeholder(tf.float32, shape=[None, 784])
```
Let's not commit commented code

goodfeli commented on examples/imagenet_featadvs/attack_model_featadv.py:
```diff
@@ -0,0 +1,160 @@
+        adv_x = ffg.generate(x_image, g, **ffg_params)
+    elif FLAGS.attack_type == 'ffg_untarg':
+        print("FFG")
+        # ['h_conv1', 'h_conv2', 'h_fc1', 'h_pool1', 'h_pool2', 'h_pool2_flat',
```
More dead code

fartashf replied:
Fixed in 47b04aa.

goodfeli commented on examples/imagenet_featadvs/attack_model_featadv.py:
```diff
@@ -0,0 +1,160 @@
+        " 'bim'->'basic iterative method'")
+    flags.DEFINE_string('checkpoint_dir', '',  # default_checkpoint_dir,
+                        'Checkpoint directory to load')
+    # ['h_conv1', 'h_conv2', 'h_fc1', 'h_pool1', 'h_pool2', 'h_pool2_flat',
```
Dead code

fartashf replied:
Fixed in 47b04aa.

goodfeli commented:
It looks like there aren't any tests, can you add at least one test per new function / class?

goodfeli commented on cleverhans/feat_advs.py:
```diff
@@ -0,0 +1,290 @@
+def ffg(x, s_feat, g_feat, eps=0.3, ord=np.inf,
+        clip_min=None, clip_max=None, targeted=False):
+    """
+    TensorFlow implementation of the Fast Gradient Method.
```
This docstring seems wrong; presumably this isn't the Fast Gradient Method

fartashf replied:
Fixed in 7f5b62d and 9459f33.

goodfeli commented on cleverhans/feat_advs.py:
```diff
@@ -0,0 +1,290 @@
+import numpy as np
```
Module level docstring

goodfeli replied:
You should put a link to the paper in here somewhere, or I think most people won't know what this is.

Also, for your sake, you probably want to give people instructions for citing the paper :)

fartashf replied:
Docstring added in 4ec1f53. Regarding citation, is it enough to include the link to arxiv?

npapernot replied:
Yes it should be fine to add the link, at least this is what we have done so far with other attacks :)

yenchenlin commented on cleverhans/feat_advs.py:
```diff
@@ -0,0 +1,290 @@
+
+    def __init__(self, model, back='tf', sess=None):
+        """
+        Create a FastGradientMethod instance.
```
This should be `Create a FastFeatureGradient instance`, right?

fartashf replied:
Fixed in c80b238. Thank you.

fartashf commented:
Yeah. Thanks. This PR is unfinished. I'm resolving the issues.

On Thu, Oct 5, 2017 at 12:33 PM, Yen <notifications@github.com> wrote:

> *@yenchenlin* commented on this pull request.
> ------------------------------
>
> In cleverhans/feat_advs.py
> <https://github.com/tensorflow/cleverhans/pull/283#discussion_r142990386>:
>
> > +    # If clipping is needed, reset all values outside of [clip_min, clip_max]
> +    if (clip_min is not None) and (clip_max is not None):
> +        adv_x = tf.clip_by_value(adv_x, clip_min, clip_max)
> +
> +    return adv_x
> +
> +
> +class FastFeatureGradient(Attack):
> +
> +    """
> +    This is similar to FGSM but done for the internal representations.
> +    """
> +
> +    def __init__(self, model, back='tf', sess=None):
> +        """
> +        Create a FastGradientMethod instance.
>
> This should be Create a FastFeatureGradient instance, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/cleverhans/pull/283#pullrequestreview-67436751>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG-A-c3Pn5yyR2bUEkkbMFJ6XE879lhfks5spQThgaJpZM4PlLap>
> .
>

yenchenlin commented on cleverhans/feat_advs.py:
```diff
@@ -0,0 +1,265 @@
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+
+        def clip_eta(eta):
```
I think you can use `clip_eta` function defined in `utils_tf.py` [here](https://github.com/tensorflow/cleverhans/blob/5114ebd6b54c4197819042a7d716b2ac4e8285a0/cleverhans/utils_tf.py#L396-L421).

fartashf replied:
Fixed in b73717bdc1206508d1e425fd84a5b1298ba59705.

fartashf commented:
Thanks for the comments. I have addressed the comments except for the one regarding writing tests. The example in `imagenet_featadvs/` is supposed to be a very basic test which is not converging to 0 right now. Let me know if you have any ideas for unit tests and if you found any bugs through reviewing.

goodfeli commented:
The tests are failing with "from utils_tf import clip_eta" getting an ImportError

npapernot commented on cleverhans/feat_advs.py:
```diff
@@ -0,0 +1,151 @@
+        loss = -tf.reduce_sum(tf.square(a_feat - g_feat), axis)
+
+        # Define gradient of loss wrt input
+        grad, = tf.gradients(loss, x)
```
Shouldn't this be `adv_x` here instead of `x`?

fartashf replied:
Thanks for catching that. Fixed in 11919a2.

npapernot commented on cleverhans/feat_advs.py:
```diff
@@ -0,0 +1,151 @@
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise ValueError("Norm order must be either np.inf, 1, or 2.")
+        if self.back == 'th':
```
You can remove this given that we dropped Theano support.

fartashf replied:
Removed in 30b4472.

npapernot commented:
@fartashf are you able to take a look at this PR in the next couple of days?

fartashf commented:
@npapernot thanks for the reminder. I fixed the issues and added a test. The tests fail for test_mnist_tutorial_tf.TestMNISTTutorialTF. I guess it's again due to a tight threshold.

goodfeli commented:
Quoting the failure here:
    Traceback (most recent call last):
      File "/home/travis/build/tensorflow/cleverhans/tests_tf/test_mnist_tutorial_tf.py", line 27, in test_mnist_tutorial_tf
        self.assertLess(report.train_clean_train_adv_eval, 0.036)
    AssertionError: 0.036799999999999999 not less than 0.036

goodfeli commented on tests_tf/test_attacks.py:
```diff
@@ -777,5 +778,82 @@ def test_multiple_initial_random_step(self):
+        super(TestFastFeatureAdversaries, self).setUp()
+        import tensorflow as tf
+
+        def make_imagenet_cnn(input_shape=(None, 224, 224, 3)):
```
How long does this test take to run? The model looks pretty big for a Travis test

goodfeli replied:
19 seconds isn't so bad. Is there anything trivial that can speed it up? For example, it looks like the batch size is 100. Does the test still work with batch size 1?

Does the attack still succeed on unrealistically small models, e.g. with only 32 filters per layer?

fartashf replied:
Changed 100 to 10 in 58a7054 and it's faster now. Thanks for the suggestion.

goodfeli commented:
Since this changes `FastGradientMethod` it does seem possible the failure is real.

Notes for my future self / other people following the thread: this PR changes `FastGradientMethod` to call `clip_eta` rather than having its own version of the `clip_eta` code internally. This is a great refactor (thanks Fartash!) but I'm concerned that there's something subtly different between the two versions that we don't know about.

Fartash: are the tests failing on your machine too or just on Travis? It looks like they fail only for newer version of TensorFlow. Do they fail consistently or flakily? From one run of Travis it looks like we have 4 trials that all support the hypothesis that the tests fail consistently for TensorFlow > 1.1. Since they consistently passed before, this looks like something did meaningfully change in `clip_eta`.

fartashf commented:
My bad for mixing multiple PRs. This PR is actually not changing the `FastGradientMethod`. It only refactors `BasicIterativeMethod` to use `clip_eta`. This should not affect the tutorial. `clip_eta` was introduced in 47d85a6d73401d60f21be0653bf7bae78b4c459b and later was moved to `utils_tf.py` in 766add490bef3fc48393b4056973f62ace3e1f55.

Tests don't fail on my machine. I tested with Tensorflow 1.4.0 using python 2.7.14 and 3.6.3.

goodfeli commented:
Are you sure the test only takes 19 seconds? The tests for this PR took up to 33 minutes to run but if I look at another currently open PR they're consistently under 26.

https://travis-ci.org/tensorflow/cleverhans/builds/314515713?utm_source=github_status&utm_medium=notification

goodfeli commented:
This one is also just barely over 26 for only one of 4 runs: https://travis-ci.org/tensorflow/cleverhans/builds/314921361?utm_source=github_status&utm_medium=notification

goodfeli commented:
Thanks, it does look a lot faster now.

Unfortunately, we now have two more examples of the tensorflow 1.1 tests failing, and by exactly the same amount of error as before:

Traceback (most recent call last):
  File "/home/travis/build/tensorflow/cleverhans/tests_tf/test_mnist_tutorial_tf.py", line 27, in test_mnist_tutorial_tf
    self.assertLess(report.train_clean_train_adv_eval, 0.036)
AssertionError: 0.036799999999999999 not less than 0.036

Since this isn't happening to pull requests in other branches, I do think there's something unusual about this branch.

One thing you might try is to rebase your branch on top of master and run the tests locally. If the rebase gets the tests to fail locally, that would give you some more convenient signal for debugging. You'll need to push -f back to the pull request after the merge.

fartashf commented:
I rebased and the tests still pass on my machine.
To make the PR simpler, I separated the part regarding `clip_eta` into a new PR #326. We will get a clue after that one is tested.

goodfeli commented:
Thanks! Feel free to push -f the remainder of the code to this branch. If we're lucky, it will just pass.

fartashf commented:
The checks have passed. I think it is only because of randomness in the tests. The only change I made was moving part of the code to PR #326.

goodfeli commented:
Thanks!
